Add an eta-v analysis page for the source query

ajstewart opened the pull request:
@marxide I wanted to practise some python and Django stuff so I thought I'd have a go at making something for #410.

I've somewhat cobbled together some sort of solution that seems to work. How good it is I'm not sure. While it's snappy locally I'm concerned at how it would run on a full run on Nimbus though it should still be fast. I've also added png previews of measurements on the light curve hovertool - I wonder if that is faster to gather than the fits...

This is not complete but I wondered what you thought. There are a few warnings in the console but I cannot figure them out, my front-end skills weren't quite there!

Easier to show in a video, though I couldn't find a way to quickly save it at double speed to I suggest scrubbing 😅

https://user-images.githubusercontent.com/3306161/143328350-039658ec-c7c4-491f-a0ab-2600d41d20b2.mov

Fixes #410.

ajstewart commented:
I've noticed the zooming on the bokeh plot doesn't work in the low value regions. Not sure why as of yet.

marxide commented on vast_pipeline/plots.py:
```diff
@@ -261,3 +304,289 @@ def plot_lightcurve(
+    Returns:
+        Bokeh grid object containing figure.
+    """
+    df = pd.DataFrame(source.values())
```
Selecting only the columns we need will help performance

```suggestion
    df = pd.DataFrame(source.values("id", "eta_peak", "eta_int", "v_peak", "v_int", "n_meas_sel"))
```

ajstewart commented on templates/base.html:
```diff
@@ -214,63 +214,84 @@
+              </div>
+            {% endif %}
+          </ul>
+        {% endblock messages %}
```
Only change here is formalising the block.

ajstewart commented on vast_pipeline/pipeline/association.py:
```diff
@@ -1251,7 +1251,7 @@ def association(images_df: pd.DataFrame, limit: Angle, dr_limit: float,
         ra_wrap_mask = sources_df.ra <= 0.1
         sources_df['ra_wrap'] = sources_df.ra.values
-        sources_df.at[
+        sources_df.loc[
```
This and the one below solved the first testing failure with the updated dependency (probably pandas). `at` shouldn't be used to get a selection of parameters.

ajstewart commented on vast_pipeline/pipeline/finalise.py:
```diff
@@ -51,18 +51,25 @@ def calculate_measurement_pair_aggregate_metrics(
-        .reset_index(drop=True)  # keep only the row indices
-    ][[f"vs_{flux_type}", f"m_{flux_type}"]]
+    check_df = measurement_pairs_df.query(f"abs(vs_{flux_type}) >= @min_vs")
+
+    if check_df.empty:
+        pair_agg_metrics = pd.DataFrame(
+            columns=[f"vs_{flux_type}", f"m_{flux_type}", "source"]
+        )
+    else:
+        pair_agg_metrics = measurement_pairs_df.iloc[
+            check_df
+            .groupby("source")
+            .agg(m_abs_max_idx=(f"m_{flux_type}", lambda x: x.abs().idxmax()),)
+            .astype(np.int32)["m_abs_max_idx"]  # cast row indices to int and select them
+            .reset_index(drop=True)  # keep only the row indices
+        ][[f"vs_{flux_type}", f"m_{flux_type}", "source"]]
 
     pair_agg_metrics = pair_agg_metrics.abs().rename(columns={
         f"vs_{flux_type}": f"vs_abs_significant_max_{flux_type}",
         f"m_{flux_type}": f"m_abs_significant_max_{flux_type}",
-    })
+    }).set_index('source')
```
This is the second fix to the updated dependancy. For some reason in the basic association test, while there are no sources that meet the minimum `vs` the code was still attempting to do the group by lambda idmax calculation on an empty frame - which failed in the GitHub tests (but not locally for me).

I've covered this by splitting this calculation a bit and only going ahead when it is not empty.

Please check this bit that you're happy with it. I also moved the setting the source as index as I was suspicious of doing it early as this could have caused duplicated index values which I'm always cautious about.

ajstewart commented on docs/exploringwebsite/sourceanalysis.md:
```diff
@@ -0,0 +1,71 @@
+![!Source light curve and information.](../img/analysis-page-source-display.png){: loading=lazy }
+
+The source information panel contains a link to go to the full [source detail page](sourcedetail.md) and the ability to [favourite a source](sourcetagsfavs.md) directly from this page.
+Note that as of the current version, adding a source to favourites will reset the analysis page.
```
I plan to fix this in a different PR.

ajstewart commented:
@marxide I think this is now ready. This turned into the plot and fixing the current test issues with 3.8 and 3.9 as I merged it...

To summarise, main new features and fixes are:
* The whole eta-v plot interface + doc pages.
  * Now with added Datashader for high source counts.
* Cutouts on the light curve hovertools.
* Bumped minimum Django and Dask, bumped and fixed bokeh version.
* Full new `poetry.lock` generated on 3.8.12.
* Test python versions updated.
* Two fixes for regressions with updating the dependencies (I suspect pandas to be the main culprit).
* Added missing docs from last PR that I noticed (about `Q_CLUSTER_MAX_ATTEMPTS`).
* Added doc badge to `README.md`.
